src/procedures/proc_shortest_path.c

```diff
@@ -0,0 +1,291 @@
+
+	double c;
+	info = GrB_Vector_extractElement_FP64(&c, d, t_id);
+	SIArray_Append(&cost, SI_DoubleVal(c));
+
+	// Determine traversed edge type.
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	Schema *schema = GraphContext_GetSchema(gc, relation, SCHEMA_EDGE);
+	int r_id = schema->id;
+
+	while(hops) {
+		/* p[t_id] contains the source node leading to parent
+		 * on the shortest path from `s` to `t`. */
+		info = GrB_Vector_extractElement_UINT64(&s_id, p, t_id);
+		assert(info == GrB_SUCCESS);
+		s_id--; // as p is 1 index based.
+
+		// Get edge connecting src to dest.
+		Graph_GetEdgesConnectingNodes(g, s_id, t_id, r_id, &edges);
+		if(array_len(edges) > 1) {
+			// TODO: handle multiple edges.
+			assert("handle multiple edges" && false);
+		}
+
+		Path_AppendEdge(path, edges[0]);
+		array_clear(edges);
+
+		Graph_GetNode(g, s_id, &n);
+		Path_AppendNode(path, n);
+
+		info = GrB_Vector_extractElement_FP64(&c, d, s_id);
+		SIArray_Append(&cost, SI_DoubleVal(c));
+
+		// Update
+		t_id = s_id;
+		hops--;
+	}
+
+	array_free(edges);
+	// Reverse path such that source node is at the beginning.
+	Path_Reverse(path);
+	SIArray_Reverse(cost);
+
+	*path_output = SI_Path(path);
+	*cost_output = cost;
+}
+
+ProcedureResult Proc_ShortestPathInvoke(ProcedureCtx *ctx, const SIValue *args) {
+	if(array_len((SIValue *)args) != 5) return PROCEDURE_ERR;
+
+	ctx->privateData = NULL;
+
+	Node *startNode = (Node *)(args[0].ptrval);
+	Node *endNode = (Node *)(args[1].ptrval);
+	const char *weightProperty = args[2].stringval;
+	const char *relationship = args[3].stringval;
+	double defaultValue = args[4].doubleval;
+
+	GrB_Matrix M = _BuildWeightMatrix(relationship, weightProperty, defaultValue);
+	if(M == GrB_NULL) return PROCEDURE_ERR;
+
+	ShortestPath_Ctx *pdata = rm_malloc(sizeof(ShortestPath_Ctx));
+	pdata->M = M;
+	pdata->endNode = endNode;
+	pdata->startNode = startNode;
+	pdata->relationship = relationship;
+	pdata->defaultValue = defaultValue;
+	pdata->weightProperty = weightProperty;
+	pdata->pathProduced = false;
+
+	pdata->output = array_new(SIValue, 4);
+	pdata->output = array_append(pdata->output, SI_ConstStringVal("path"));
+	pdata->output = array_append(pdata->output, SI_NullVal()); // Place holder.
+	pdata->output = array_append(pdata->output, SI_ConstStringVal("cost"));
+	pdata->output = array_append(pdata->output, SI_NullVal()); // Place holder.
+
+	ctx->privateData = pdata;
+	return PROCEDURE_OK;
+}
+
+SIValue *Proc_ShortestPathStep(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
+
+	SIValue *res = NULL;
+	ShortestPath_Ctx *pdata = (ShortestPath_Ctx *)ctx->privateData;
+	Node *s = pdata->startNode;
+	Node *t = pdata->endNode;
+	NodeID s_id = ENTITY_GET_ID(s);
+	NodeID t_id = ENTITY_GET_ID(t);
+	GrB_Vector d = GrB_NULL;    // Pointer to the vector of distance
+	GrB_Vector pi = GrB_NULL;   // Pointer to the vector of parent
+	GrB_Vector h = GrB_NULL;    // Pointer to the vector of hops
+
+	if(pdata->pathProduced) goto cleanup;
```

Comment:
In which scenario this could happen?

Reply:
if someone calls `Proc_ShortestPathStep` multiple times, e.g. `op_procedure_call` `_yield` function.